source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -399,6 +406,28 @@ ReadOrParseState Filter::readProxyHeader(Network::ListenerFilterBuffer& buffer)
   auto raw_slice = buffer.rawSlice();
   const char* buf = static_cast<const char*>(raw_slice.mem_);
 
+  if (config_.get()->allowRequestsWithoutProxyProtocol()) {
+
+    auto matchv2 = !memcmp(buf, PROXY_PROTO_V2_SIGNATURE,
+                           std::min<size_t>(PROXY_PROTO_V2_SIGNATURE_LEN, raw_slice.len_));
+    auto matchv1 = !memcmp(buf, PROXY_PROTO_V1_SIGNATURE,
+                           std::min<size_t>(PROXY_PROTO_V1_SIGNATURE_LEN, raw_slice.len_));
```

Comment:
Since the code of `readProxyHeader` already did the header match, I'm thinking why we didn't reuse it.

For example, the line 438 will know the it doesn't match the header both for v1 and v2 https://github.com/envoyproxy/envoy/pull/18951/files#diff-a35fc10b1ce97239d02ac1791aac05069eaf253d8efc77cefe93749f901e36abR438

Then you can check `allowRequestsWithoutProxyProtocol` right here.

Also line 489 can know when the data less than v2 header size whether match the v1 header https://github.com/envoyproxy/envoy/pull/18951/files#diff-a35fc10b1ce97239d02ac1791aac05069eaf253d8efc77cefe93749f901e36abR489

Then you can check `allowRequestsWithoutProxyProtocol` flag at there also.  If true, then we skip.

Reply:
I like the idea of reusing parsing logic we already have, but your proposal would actually break this test:
```c++
TEST_P(ProxyProtocolTest, AllowTinyNoProxyProtocol) {
  // allows a small request (less bytes than v1/v2 signature) through even though it doesn't use
  // proxy protocol
  envoy::extensions::filters::listener::proxy_protocol::v3::ProxyProtocol proto_config;
  proto_config.set_allow_requests_without_proxy_protocol(true);
  connect(true, &proto_config);

  std::string msg = "data";
  EXPECT_GT(PROXY_PROTO_V1_SIGNATURE_LEN, msg.length());
  EXPECT_GT(PROXY_PROTO_V2_SIGNATURE_LEN, msg.length());

  write(msg);

  expectData(msg);

  disconnect();
}
```
This test would infinite loop (waiting for more bytes for v1 proxy protocol [here](https://github.com/envoyproxy/envoy/blob/887d0bc55e917bec5c6b3d024275b60c1c3e5bd5/source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc#L489)) with your proposal, unless I added an additional check. Since I need the additional checks anyways, I opted to have all the logic in one place rather than spread around the function.

Reply:
With the additional check on match for proxy protocol v1, I implemented the idea behind your suggestion in https://github.com/envoyproxy/envoy/commit/a23ac5f1e2613be72526f5dcb70a76b13deef126. I'm not convinced it's better since it splits the logic for allowRequestsWithoutProxyProtocol() into two places, although I don't feel strongly so if you like this better I'm happy to keep it as is; or revert it back. Thanks!

Reply:
I reverted https://github.com/envoyproxy/envoy/pull/18951/commits/a23ac5f1e2613be72526f5dcb70a76b13deef126 in  https://github.com/envoyproxy/envoy/pull/18951/commits/6075b676d7deba6559c3eee3be4f43b02d778f86 since the latter check (where line 489 was) would need to check the early bytes for both v1 and v2 to be correct; so I think it is better to do that check once up front rather than in two places

Reply:
> I like the idea of reusing parsing logic we already have, but your proposal would actually break this test:

It probably due to this: https://github.com/envoyproxy/envoy/blob/25cc9ea48ba18930e1235dfb6a6a236d59c8b08a/test/extensions/filters/listener/proxy_protocol/proxy_protocol_test.cc#L83

It return false means close the connection when timeout.

Reply:
It's because the current proxy protocol filter is written with the assumption that we are receiving either proxy protocol v1 or v2, and that we will eventually have enough bytes to tell which one it is. Notice here that we just wait for the end of proxy protocol v1 header (rather than checking the earlier bytes for v1/v2 signature): https://github.com/envoyproxy/envoy/blob/887d0bc55e917bec5c6b3d024275b60c1c3e5bd5/source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc#L475

I could update this logic to check if the bytes we have seen so far match v1 or v2 (similar to what I added in this PR, but gated by `allowRequestsWithoutProxyProtocol()`, and similar to what we already do for v2 proxy protocol [here](https://github.com/envoyproxy/envoy/blob/887d0bc55e917bec5c6b3d024275b60c1c3e5bd5/source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc#L433)) so we could keep the `search_index_` optimization. This would require another couple state variables on the filter (to remember whether what we already saw was valid v1 or v2) but I'm happy to do so if we think the complexity is worth the potential performance improvements (I initially leaned against since the v2 signature is at most 12 bytes) 

I wanted us to gracefully handle small requests, e.g. just sending `"data"` and fast failing (since we know this is neither v1 nor v2) rather than wait for timeout and forward (which the [docs](https://github.com/envoyproxy/envoy/blob/9d5627a0879b0a029e90515137c108e1d2884bfc/api/envoy/api/v2/listener.proto#L145-L152) recommend against)

Reply:
> It's because the current proxy protocol filter is written with the assumption that we are receiving either proxy protocol v1 or v2, and that we will eventually have enough bytes to tell which one it is. Notice here that we just wait for the end of proxy protocol v1 header (rather than checking the earlier bytes for v1/v2 signature):
> 
> https://github.com/envoyproxy/envoy/blob/887d0bc55e917bec5c6b3d024275b60c1c3e5bd5/source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc#L475
> 
> I could update this logic to check if the bytes we have seen so far match v1 or v2 (similar to what I added in this PR, but gated by `allowRequestsWithoutProxyProtocol()`, and similar to what we already do for v2 proxy protocol [here](https://github.com/envoyproxy/envoy/blob/887d0bc55e917bec5c6b3d024275b60c1c3e5bd5/source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc#L433)) so we could keep the `search_index_` optimization. This would require another couple state variables on the filter (to remember whether what we already saw was valid v1 or v2) but I'm happy to do so if we think the complexity is worth the potential performance improvements (I initially leaned against since the v2 signature is at most 12 bytes)
> 

Emm... with current implement If we matched the header here, then we match here again after few lines, which doesn't  make sense to me.

For `search_index_`, I thought we just need to compare each bytes both for v1 and v2, then that will be enough? maybe I missed something.

@ggreenway WDYT?

> I wanted us to gracefully handle small requests, e.g. just sending `"data"` and fast forwarding as non-proxy protocol (since we already know this is neither v1 nor v2) rather than wait for timeout and forward (which the [docs](https://github.com/envoyproxy/envoy/blob/9d5627a0879b0a029e90515137c108e1d2884bfc/api/envoy/api/v2/listener.proto#L145-L152) recommend against)

Yes, I agree with that.

Reply:
@soulxu I believe I understand what you were suggesting (reuse existing parsing logic where possible) and implemented that in https://github.com/envoyproxy/envoy/pull/18951/commits/2cb45268cf06d44c96c68c604ecc7d585aeb9a51

Please let me know if I misunderstood your suggestion. Thanks for all the feedback!